plugins/events-backend-module-http/src/service/HttpEndpointEventPublisher.ts

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2022 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { errorHandler } from '@backstage/backend-common';
+import { Config } from '@backstage/config';
+import { EventPublisher, EventBus } from '@backstage/plugin-events-node';
+import express from 'express';
+import Router from 'express-promise-router';
+import { Logger } from 'winston';
+import { EventHttpRequestValidator } from './EventHttpRequestValidator';
+
+/**
+ * Publishes events received from their origin (e.g., webhook events from an SCM system)
+ * via HTTP endpoint and passes the request body as event payload to the registered subscribers.
+ *
+ * @public
+ */
+export class HttpEndpointEventPublisher implements EventPublisher {
+  private eventBus?: EventBus;
+
+  static fromConfig(env: {
+    config: Config;
+    logger: Logger;
+    router: express.Router;
+    validators?: Record<string, EventHttpRequestValidator>;
+  }): HttpEndpointEventPublisher {
+    const topics = env.config.getStringArray('events.modules.http.topics');
+    return new HttpEndpointEventPublisher(
+      env.logger,
+      env.router,
+      topics,
+      env.validators,
+    );
+  }
+
+  private constructor(
+    private logger: Logger,
+    router: express.Router,
+    topics: string[],
+    validators?: Record<string, EventHttpRequestValidator>,
+  ) {
+    router.use(this.createRouter(topics, validators));
+  }
+
+  async connect(eventBus: EventBus): Promise<void> {
+    this.eventBus = eventBus;
+  }
+
+  private createRouter(
+    topics: string[],
+    validators?: Record<string, EventHttpRequestValidator>,
+  ): express.Router {
+    const router = Router();
+    router.use(express.json());
+    router.use(errorHandler());
+
+    topics.forEach(topic =>
+      this.addRouteForTopic(router, topic, validators?.[topic]),
+    );
+
+    return router;
+  }
+
+  private addRouteForTopic(
+    router: express.Router,
+    topic: string,
+    validator?: EventHttpRequestValidator,
+  ): void {
+    const path = `/${topic}`;
+
+    router.post(path, async (request, response) => {
+      if (validator && !(await validator(request))) {
+        response.status(404).send();
```

Comment:
This is likely debatable, but my gut feeling is that this should be a 400 and not a 404, to have some way of differentiating between a wrong URL and something else having gone wrong.

Reply:
Also, since the sender of the webhook may expect a JSON response, it could be nice to send something along the lines of `{ status: 'validation failed' }`?

Reply:
It depends. I think we could even use Forbidden. If you consider malicious requests which you want to block, it is better to share less information.

..but maybe this can be a decision of the validator 🤔 
Then you have flexibility if you prefer "Not Found", "Forbidden" or "Bad Request" and which data you want to return if any.

Reply:
Yep, that sounds like a decent idea, letting the validator decide.

Reply:
I've adjusted it to allow to return more than just a boolean.